### DISCUSSION TITLE
Mapping abstract property

### DISCUSSION THREAD

**Discussion by gauravgupta429** [mapstruct/mapstruct#2944] [Q&A] [answered]
Hi,

Following are entity and dto details.

**Entity Details:**

public class PartyPreference{

	private Preference preference;

}
public abstract class Preference {

  private String partyType;

  private String alertMode;

   private String communicationMode;

  private String preferredLanguage;

  private String preferredCurrency;

}

public class IndividualPreference extends Preference {

  private LocalTime preferredTime;

}

public class NonIndividualPreference extends Preference {

  private LocalTime preferredTime;

}

**DTO Details:**

public class PartyPreferenceDTO{

	private PreferenceDTO preferenceDTO;

}
public abstract class PreferenceDTO {

  private String partyType;

  private String alertMode;

   private String communicationMode;

  private String preferredCurrency;

}

public class IndividualPreferenceDTO extends PreferenceDTO {

  private String preferredLanguage;

}

public class NonIndividualPreferenceDTO extends PreferenceDTO {

  private LocalTime preferredTime;

}

Please suggest how to write mapper for abstract property.

**Comment by chenzijia12300** [comment, marked as answer]
I think you can take a look at this [Subclass Mapping](https://mapstruct.org/documentation/stable/reference/html/#sub-class-mappings)

The final implementation looks like this

```
@Mapper(subclassExhaustiveStrategy = SubclassExhaustiveStrategy.RUNTIME_EXCEPTION)
public interface PreferenceMapping {

    PreferenceMapping INSTANCE = Mappers.getMapper(PreferenceMapping.class);


    @Mapping(target = "preferenceDTO",source = "preference")
    PartyPreferenceDTO toDto(PartyPreference preference);

    @SubclassMapping(source = IndividualPreference.class,target = IndividualPreferenceDTO.class)
    @SubclassMapping(source = NonIndividualPreference.class,target = NonIndividualPreferenceDTO.class)
    PreferenceDTO map(Preference preference);
}
```

**Reply by gauravgupta429** [reply]
Thanks for your help!!
Working fine now..

**Reply by chenzijia12300** [reply]
I'm glad it's enough to help you (￣▽￣)"
